### DISCUSSION TITLE
Is it possible to use BuildKit with Testcontainers in .net ?

### DISCUSSION THREAD

**Discussion by osa0805** [testcontainers/testcontainers-dotnet#1193] [Q&A] [answered]
When trying to build image from dockerfile using Testcontainers there is an error of "the * option requires BuildKit.".
Is it possible to use BuildKit somehow ?

**Comment by HofmeisterAn** [comment, marked as answer]
No. BuildKit is not part of the [Docker Engine API](https://docs.docker.com/engine/api/).

**Reply by osa0805** [reply]
How it correlates with "[BuildKit](https://github.com/moby/buildkit) is an improved backend to replace the legacy builder. BuildKit is the default builder for users on Docker Desktop, and Docker Engine as of version 23.0." 
https://docs.docker.com/build/buildkit/ 

Can you elaborate more please?

**Reply by HofmeisterAn** [reply]
There is not much documentation available. It was not supported in the past. Maybe Docker added support recently (https://github.com/docker/for-linux/issues/1136#issuecomment-2159656470), but I'm not sure. If it was added, we likely need to update [https://github.com/dotnet/Docker.DotNet](https://github.com/dotnet/Docker.DotNet) first.

**Reply by osa0805** [reply]
So it is possible that you will create some Task/Issue to check it ? :)

**Reply by HofmeisterAn** [reply]
@kiview do you have more info on this? If it is already (now) supported, just give me a heads-up. Then we can turn this into an issue and maybe try to support/implement it someday.

**Reply by rgl** [reply]
FWIW, not having support for this will make things harder, as the newer dockerfile syntax is much user-friendly than the old one.

It seems https://github.com/moby/moby/issues/41711#issuecomment-2160179739 has a solution, not sure whether it should be tried in this repo or in https://github.com/dotnet/Docker.DotNet thou. Any idea?

**Reply by HofmeisterAn** [reply]
Since it looks like BuildKit is enabled/disabled per API (`/build`) call, as mentioned, I think we need to update Docker.DotNet first and include the new field (`version`) in the [ImageBuildParameters](https://github.com/dotnet/Docker.DotNet/blob/aacf3c26131f582ca8acc34084663a4b79e28d38/tools/specgen/specgen.go#L174-L176) type (updating Docker.DotNet to the latest Docker Engine API in general). Otherwise, we cannot set the version to 2.

**Reply by osa0805** [reply]
Who is responsible for Docker.DoNet ? Should I make an issue there or something like that ?

**Reply by kiview** [reply]
@osa0805 That would definitely make sense. I'll also ask internally at Docker, whats the situation with the BuildKit support in the API.

**Reply by kiview** [reply]
@HofmeisterAn @osa0805 I talked internally with colleagues at Docker, and the BuildKit API is not officially publicly exposed for integration. Specifically, although there exist some API endpoint, it requires significant client-side code/logic and not suitable for simple public integration. So I wouldn't expect us to be able to provide BuildKit support through API integration any time soon, but will keep the internal discussion open.

**Reply by osa0805** [reply]
I am using this feature a lot in my work and it is great!
Because of the missing BuildKit support I was forced to make separate Dockefile without BuildKit instructions and maintain two separate versions. 
It will be a pity if the BuildKit integration will not be possible :(

**Reply by HofmeisterAn** [reply]
You can use the CLI to build the image and pass the image name as an argument to `dotnet test`. I know it’s not the smoothest way to build the image and run tests automatically one after the other, but maybe you can find a decent workflow until the Docker Engine API fully supports it.

Also, you could check out the [.NET SDK Container Building Tools](https://github.com/dotnet/sdk-container-builds?tab=readme-ov-file). They let you automatically build your image from a project file. You could pair it with the after-build target so your tests always have access to the latest built version (I am experimenting with this approach right now).

**Reply by bjuraga** [reply]
Is this still true? No support for buildkit?

**Reply by HofmeisterAn** [reply]
Why the thumbs-down 😆? This isn't something we (the Testcontainers maintainers) can control. Yes, it's still unsupported, and it probably won't change anytime soon.

This limitation comes from the Docker Engine API itself. We'd need Docker to add support for it (move it from the client to the backend).

If this is important to you, the best next step is to share that feedback directly with Docker.

**Reply by bjuraga** [reply]
The thumbs down is for passing down the issue. 

For example, outside of testcontainers i am able to do something like set the env var to enable buildkit and inline call docker build ...

So, from a testcontainer user point of view it seems trivial to add something like env vars support before running a command. Could even be possible, seems like a simple thing to  do and could be useful just as an overall API to have WithEnvironmentVariable().
From the web:

```
To set the BuildKit environment variable when running the docker build command, run: $ DOCKER_BUILDKIT=1 docker build .
```

**Reply by HofmeisterAn** [reply]
This is probably a misunderstanding. Testcontainers uses the [Docker Engine API](https://docs.docker.com/reference/api/engine/version/v1.52/), which has nothing to do with the Docker CLI or anything the CLI supports. They're two separate things. Testcontainers is **not** a CLI wrapper.

**Reply by bjuraga** [reply]
In that case i rewoke my thumbs down and offer my sincere apology.

**Reply by HofmeisterAn** [reply]
No worries! It makes sense to think Testcontainers (could) use the CLI, but that's not how it works and would likely block many features.

Building images without BuildKit is annoying, I'll add support as soon as it's possible (or someone comes up with a workaround).

**Reply by bjuraga** [reply]
I can come up with a workaround to treat the image building differently than the rest of the containers. Is there a hard reason to not wrap the docker cli for the image? Could be an explicit second way of building images with clearly stated no support from your side.
